### PR TITLE
Change users models

### LINKED ISSUE #61: Create tests for users
Add unit tests for users

### PR DESCRIPTION
**Pull request by AvivLiberman**
Fixes #61 

Django default user already has email, first_name, last_name.
I removed those 3 fields from our User and .

## Changes you made
- removed those 3 fields from our User
- created a static method that will create our user with all required fields
- added tests for creating team and user

from now on we must write tests for each PR

### CODE REVIEW

File: tasks/models.py [hunk 1]
```diff
@@ -19,8 +19,10 @@ class Priority(ChoiceEnum):
 class Task(models.Model):
     title = models.CharField(max_length=200)
     created_date = models.DateTimeField(auto_now_add=True)
-    assignee = models.ForeignKey(User, on_delete=models.RESTRICT, related_name='assigneeTasks')
-    created_by = models.ForeignKey(User, on_delete=models.RESTRICT, related_name='createdTasks')
+    assignee = models.ForeignKey(
+        User, on_delete=models.RESTRICT, related_name='assigneeTasks')
+    created_by = models.ForeignKey(
+        User, on_delete=models.RESTRICT, related_name='createdTasks')
```
Comment (orzionpour) [inline tasks/models.py:25]: flake8 fix?

File: tasks/models.py [hunk 2]
```diff
@@ -30,8 +32,9 @@ def __str__(self) -> str:
 
 
 class Comment(models.Model):
-    user_id = models.ForeignKey(User, on_delete=models.RESTRICT, related_name='comments')
-    task_id = models.ForeignKey(Task, on_delete=models.CASCADE)
+    user = models.ForeignKey(
+        User, on_delete=models.RESTRICT, related_name='comments')
+    task = models.ForeignKey(Task, on_delete=models.CASCADE)
```
Comment (orzionpour) [inline tasks/models.py:37]: Just another naming or there is a specific reason for the name change?

**Review by orzionpour — COMMENTED** [review]
Can you remove the dummy test? we don't need it anymore since there are real tests now.

**Comment by orzionpour** [comment]
As for the test and functions:
Django has more functions like delete user, change password, etc.. Since we are using another users table we should implement another function that updates both tables. For example, our delete function should call django's delete so we can do it directly from our table. Correct me if i go wrong here
In addition, I guess we need change_team and change_role functions.

**Comment by AvivLiberman** [comment]
I don't want that PR to be big. Any other ideas you have to add more functionality open an issue about it

**Review by Asafh7 — APPROVED** [review]

File: tasks/models.py (hunk 2 above)
Comment (AvivLiberman) [reply tasks/models.py:37]: for every foreign key django already adding _id in the end, so better to remove it now

File: tasks/models.py (hunk 1 above)
Comment (AvivLiberman) [reply tasks/models.py:25]: yes

**Review by aradz1997 — APPROVED** [review]

**Review by orzionpour — APPROVED** [review]

File: tasks/models.py (hunk 2 above)
Comment (aradz1997) [reply tasks/models.py:37]: plz add a " related_name='comments' " to the field - task
task = models.ForeignKey(Task, on_delete=models.CASCADE, related_name='comments')
@AvivLiberman

**Review by aradz1997 — CHANGES_REQUESTED** [review]

File: tasks/models.py (hunk 2 above)
Comment (AvivLiberman) [reply tasks/models.py]: Done

**Review by Asafh7 — DISMISSED** [review]

**Review by orzionpour — DISMISSED** [review]

**Review by aradz1997 — APPROVED** [review]

**Comment by natifridman** [comment]
Why `conftest.py` need to be in the main folder?

File: conftest.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import pytest
+from users.models import Role, User, Team
+
+
+@pytest.fixture
+def new_team_factory(db):
```
Comment (kobihk) [inline conftest.py:6]: I think you can do it like:
```suggestion
def team_1(db):
```
and get rid of the team_1 fixture in lines # 15-17

File: conftest.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    def create_team(name="Team1",
+                    description="very good team"):
+        team = Team.objects.create(name=name,
+                                   description=description)
+        return team
+    return create_team
```
Comment (kobihk) [inline conftest.py:12]:
```suggestion
        return Team.objects.create(name="Team1", description="very good team")
```

File: conftest.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                    team=new_team_factory("Team1", "Best team")):
+        user = User.create_user(username=username,
+                                email=email,
+                                password=password,
+                                first_name=first_name,
+                                last_name=last_name,
+                                role=role,
+                                team=team)
+        return user
+    return create_user
```
Comment (kobihk) [inline conftest.py:37]: I think you can do it something like:
```suggestion
        user1 = User.create_user(
            username="user1",
            email="user1@redhat.com",
            password="password",
            first_name="first_name",
            last_name="last_name",
            role=Role.EMPLOYEE,
            team=new_team_factory
                )
        return user1
```

File: tasks/models.py [hunk 6]
```diff
@@ -30,8 +32,10 @@ def __str__(self) -> str:
 
 
 class Comment(models.Model):
-    user_id = models.ForeignKey(User, on_delete=models.RESTRICT, related_name='comments')
-    task_id = models.ForeignKey(Task, on_delete=models.CASCADE)
+    user = models.ForeignKey(
+        User, on_delete=models.RESTRICT, related_name='comments')
```
Comment (kobihk) [inline tasks/models.py:36]: maybe
```suggestion
        User, on_delete=models.RESTRICT, related_name='user')
```

File: tasks/models.py [hunk 7]
```diff
@@ -30,8 +32,10 @@ def __str__(self) -> str:
...
+    task = models.ForeignKey(
+        Task, on_delete=models.CASCADE,  related_name='comments')
```
Comment (kobihk) [inline tasks/models.py:38]: maybe
```suggestion
        Task, on_delete=models.CASCADE,  related_name='task')
```

File: users/migrations/0001_initial.py [hunk 8]
```diff
@@ -27,13 +27,10 @@ class Migration(migrations.Migration):
...
                  models.OneToOneField(on_delete=django.db.models.deletion.CASCADE, primary_key=True, serialize=False,
                                       to='auth.user')),
-                ('email', models.EmailField(max_length=255)),
-                ('first_name', models.CharField(max_length=255)),
-                ('last_name', models.CharField(max_length=255)),
                 ('role', enumchoicefield.fields.EnumChoiceField(default=users.models.Role.EMPLOYEE,
                                                                 enum_class=users.models.Role, max_length=1)),
-                ('team_id', models.ForeignKey(on_delete=django.db.models.deletion.RESTRICT, related_name='users',
-                                              to='users.team')),
+                ('team', models.ForeignKey(on_delete=django.db.models.deletion.RESTRICT, related_name='users',
```
Comment (kobihk) [inline users/migrations/0001_initial.py:32]: maybe
```suggestion
                ('team', models.ForeignKey(on_delete=django.db.models.deletion.RESTRICT, related_name='team',
```

File: users/models.py [hunk 9]
```diff
@@ -21,12 +22,22 @@ def __str__(self) -> str:
 class User(models.Model):
     user = models.OneToOneField(
         settings.AUTH_USER_MODEL, on_delete=models.CASCADE, primary_key=True)
-    email = models.EmailField(max_length=255)
-    first_name = models.CharField(max_length=255)
-    last_name = models.CharField(max_length=255)
     role = EnumChoiceField(Role, default=Role.EMPLOYEE, max_length=1)
-    team_id = models.ForeignKey(
+    team = models.ForeignKey(
         Team, on_delete=models.RESTRICT, related_name='users')
```
Comment (kobihk) [inline users/models.py:27]:
```suggestion
        Team, on_delete=models.RESTRICT, related_name='team')
```

File: users/tests.py [hunk 10]
```diff
@@ -1,3 +1,11 @@
...
-# Create your tests here.
+
+def test_create_user(employee_1):
+    assert isinstance(employee_1, User)
+    assert employee_1.role == Role.EMPLOYEE
+
+
+def test_create_team(team_1):
+    assert isinstance(team_1, Team)
+    assert team_1.name == "Team1"
```
Comment (kobihk) [inline users/tests.py:11]: please add negative tests like: 
 - test_create_user_without_first_name(..)
 - test_create_user_with_long_name(...)
 - ...

**Review by kobihk — CHANGES_REQUESTED** [review]

File: conftest.py (hunk 3 above)
Comment (AvivLiberman) [reply conftest.py:6]: why?? I dont agree, you are losing the whole point with creating factory

File: tasks/models.py (hunk 6 above)
Comment (AvivLiberman) [reply tasks/models.py:36]: why?

File: tasks/models.py (hunk 7 above)
Comment (AvivLiberman) [reply tasks/models.py:38]: why?

File: users/migrations/0001_initial.py (hunk 8 above)
Comment (AvivLiberman) [reply users/migrations/0001_initial.py:32]: why?

File: tasks/models.py (hunk 6 above)
Comment (aradz1997) [reply tasks/models.py:36]: it should stay  "comments" im my opinion , cause when I will need to pull all the comments of a user I will write user.comments, and not user.user

**Review by aradz1997 — APPROVED** [review]

**Comment by AvivLiberman** [comment]
> Why `conftest.py` need to be in the main folder?

you are right - I removed it

**Review by Asafh7 — DISMISSED** [review]

**Review by orzionpour — APPROVED** [review]

**Review by orzionpour — APPROVED** [review]

**Review by orzionpour — APPROVED** [review]

**Review by Asafh7 — APPROVED** [review]

**Review by natifridman — APPROVED** [review]

**Pull request merged**